File: strategy.py
import pandas as pd
import config


def calculate_emas(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()
    df["ema_fast"] = df["close"].ewm(span=config.EMA_FAST, adjust=False).mean()
    df["ema_slow"] = df["close"].ewm(span=config.EMA_SLOW, adjust=False).mean()
    return df


def generate_signals(df: pd.DataFrame) -> pd.DataFrame:
    df = calculate_emas(df)

    # 1 = fast crosses above slow (golden cross = BUY)
    # -1 = fast crosses below slow (death cross = SELL)
    df["signal"] = 0
    df["prev_fast"] = df["ema_fast"].shift(1)
    df["prev_slow"] = df["ema_slow"].shift(1)

    golden_cross = (df["ema_fast"] > df["ema_slow"]) & (df["prev_fast"] <= df["prev_slow"])
    death_cross = (df["ema_fast"] < df["ema_slow"]) & (df["prev_fast"] >= df["prev_slow"])

    df.loc[golden_cross, "signal"] = 1
    df.loc[death_cross, "signal"] = -1

    df.drop(columns=["prev_fast", "prev_slow"], inplace=True)
    return df


def get_latest_signal(df: pd.DataFrame) -> dict:
    df = generate_signals(df)
    last = df.iloc[-1]
    prev = df.iloc[-2]

    return {
        "signal": int(last["signal"]),
        "close": float(last["close"]),
        "ema_fast": float(last["ema_fast"]),
        "ema_slow": float(last["ema_slow"]),
        "trend": "bullish" if last["ema_fast"] > last["ema_slow"] else "bearish",
        "prev_signal": int(prev["signal"]),
    }

File: web/src/lib/strategy.js
import { EMA_FAST, EMA_SLOW } from "./config.js";

function ema(values, span) {
  const k = 2 / (span + 1);
  const out = new Array(values.length);
  out[0] = values[0];
  for (let i = 1; i < values.length; i++) {
    out[i] = values[i] * k + out[i - 1] * (1 - k);
  }
  return out;
}

export function calculateEMAs(candles) {
  const closes = candles.map((c) => c.close);
  const emaFast = ema(closes, EMA_FAST);
  const emaSlow = ema(closes, EMA_SLOW);
  return candles.map((c, i) => ({ ...c, emaFast: emaFast[i], emaSlow: emaSlow[i] }));
}

export function generateSignals(candles) {
  const withEmas = calculateEMAs(candles);
  return withEmas.map((c, i) => {
    if (i === 0) return { ...c, signal: 0 };
    const prev = withEmas[i - 1];
    let signal = 0;
    const goldenCross = c.emaFast > c.emaSlow && prev.emaFast <= prev.emaSlow;
    const deathCross = c.emaFast < c.emaSlow && prev.emaFast >= prev.emaSlow;
    if (goldenCross) signal = 1;
    else if (deathCross) signal = -1;
    return { ...c, signal };
  });
}

export function getLatestSignal(candles) {
  const withSignals = generateSignals(candles);
  const last = withSignals[withSignals.length - 1];
  const prev = withSignals[withSignals.length - 2];
  return {
    signal: last.signal,
    close: last.close,
    emaFast: last.emaFast,
    emaSlow: last.emaSlow,
    trend: last.emaFast > last.emaSlow ? "bullish" : "bearish",
    prevSignal: prev ? prev.signal : 0,
  };
}
